From 1064fee7c23a817df966c3466b275f742f03ab2b Mon Sep 17 00:00:00 2001
From: Pantelis Antoniou <pantelis.antoniou@konsulko.com>
Date: Mon, 5 Aug 2013 22:44:09 +0300
Subject: [PATCH 08/26] arm: omap: Proper cleanups for omap_device

Removing any omap device always resulted in a crash; turns out
BUS_NOTIFY_DEL_DEVICE is not the last notifier event sent in the
course of removing the device, the correct event is
BUS_NOTIFY_UNBOUND_DRIVER, which still is not the right place to
perform the cleanup.

Signed-off-by: Pantelis Antoniou <pantelis.antoniou@konsulko.com>
---
 arch/arm/mach-omap2/omap_device.c | 29 ++++++++++++++++++++++++++---
 1 file changed, 26 insertions(+), 3 deletions(-)

diff --git a/arch/arm/mach-omap2/omap_device.c b/arch/arm/mach-omap2/omap_device.c
index 5eab056..de0d9b2 100644
--- a/arch/arm/mach-omap2/omap_device.c
+++ b/arch/arm/mach-omap2/omap_device.c
@@ -211,6 +211,26 @@ odbfd_exit:
 	return ret;
 }
 
+static void _omap_device_cleanup(struct omap_device *od)
+{
+	struct omap_hwmod *oh;
+	int i;
+
+	for (i = 0; i < od->hwmods_cnt; i++) {
+
+		oh = od->hwmods[i];
+
+		/* shutdown hwmods */
+		omap_hwmod_shutdown(oh);
+
+		/* we don't remove clocks cause there's no API to do so */
+		/* no harm done, since they will not be created next time */
+	}
+
+	/* cleanup the structure now */
+	omap_device_delete(od);
+}
+
 static int _omap_device_notifier_call(struct notifier_block *nb,
 				      unsigned long event, void *dev)
 {
@@ -218,9 +238,12 @@ static int _omap_device_notifier_call(struct notifier_block *nb,
 	struct omap_device *od;
 
 	switch (event) {
-	case BUS_NOTIFY_DEL_DEVICE:
-		if (pdev->archdata.od)
-			omap_device_delete(pdev->archdata.od);
+	case BUS_NOTIFY_UNBOUND_DRIVER:
+		/* NOTIFY_DEL_DEVICE is not the right call... */
+		od = pdev->archdata.od;
+		pdev->archdata.od = NULL;
+		if (od)
+			_omap_device_cleanup(od);
 		break;
 	case BUS_NOTIFY_ADD_DEVICE:
 		if (pdev->dev.of_node)
-- 
2.1.4

